// boxen/index.js
'use strict';
const stringWidth = require('string-width');
const chalk = require('chalk');
const widestLine = require('widest-line');
const cliBoxes = require('cli-boxes');
const camelCase = require('camelcase');
const ansiAlign = require('ansi-align');
const termSize = require('term-size');

const getObject = detail => ***REMOVED***
	let object;

	if (typeof detail === 'number') ***REMOVED***
		object = ***REMOVED***
			top: detail,
			right: detail * 3,
			bottom: detail,
			left: detail * 3
		***REMOVED***;
	***REMOVED*** else ***REMOVED***
		object = ***REMOVED***
			top: 0,
			right: 0,
			bottom: 0,
			left: 0,
			...detail
		***REMOVED***;
	***REMOVED***

	return object;
***REMOVED***;

const getBorderChars = borderStyle => ***REMOVED***
	const sides = [
		'topLeft',
		'topRight',
		'bottomRight',
		'bottomLeft',
		'vertical',
		'horizontal'
	];

	let chararacters;

	if (typeof borderStyle === 'string') ***REMOVED***
		chararacters = cliBoxes[borderStyle];

		if (!chararacters) ***REMOVED***
			throw new TypeError(`Invalid border style: $***REMOVED***borderStyle***REMOVED***`);
		***REMOVED***
	***REMOVED*** else ***REMOVED***
		for (const side of sides) ***REMOVED***
			if (!borderStyle[side] || typeof borderStyle[side] !== 'string') ***REMOVED***
				throw new TypeError(`Invalid border style: $***REMOVED***side***REMOVED***`);
			***REMOVED***
		***REMOVED***

		chararacters = borderStyle;
	***REMOVED***

	return chararacters;
***REMOVED***;

const isHex = color => color.match(/^#[0-f]***REMOVED***3***REMOVED***(?:[0-f]***REMOVED***3***REMOVED***)?$/i);
const isColorValid = color => typeof color === 'string' && ((chalk[color]) || isHex(color));
const getColorFn = color => isHex(color) ? chalk.hex(color) : chalk[color];
const getBGColorFn = color => isHex(color) ? chalk.bgHex(color) : chalk[camelCase(['bg', color])];

module.exports = (text, options) => ***REMOVED***
	options = ***REMOVED***
		padding: 0,
		borderStyle: 'single',
		dimBorder: false,
		align: 'left',
		float: 'left',
		...options
	***REMOVED***;

	if (options.borderColor && !isColorValid(options.borderColor)) ***REMOVED***
		throw new Error(`$***REMOVED***options.borderColor***REMOVED*** is not a valid borderColor`);
	***REMOVED***

	if (options.backgroundColor && !isColorValid(options.backgroundColor)) ***REMOVED***
		throw new Error(`$***REMOVED***options.backgroundColor***REMOVED*** is not a valid backgroundColor`);
	***REMOVED***

	const chars = getBorderChars(options.borderStyle);
	const padding = getObject(options.padding);
	const margin = getObject(options.margin);

	const colorizeBorder = border => ***REMOVED***
		const newBorder = options.borderColor ? getColorFn(options.borderColor)(border) : border;
		return options.dimBorder ? chalk.dim(newBorder) : newBorder;
	***REMOVED***;

	const colorizeContent = content => options.backgroundColor ? getBGColorFn(options.backgroundColor)(content) : content;

	text = ansiAlign(text, ***REMOVED***align: options.align***REMOVED***);

	const NL = '\n';
	const PAD = ' ';

	let lines = text.split(NL);

	if (padding.top > 0) ***REMOVED***
		lines = new Array(padding.top).fill('').concat(lines);
	***REMOVED***

	if (padding.bottom > 0) ***REMOVED***
		lines = lines.concat(new Array(padding.bottom).fill(''));
	***REMOVED***

	const contentWidth = widestLine(text) + padding.left + padding.right;
	const paddingLeft = PAD.repeat(padding.left);
	const ***REMOVED***columns***REMOVED*** = termSize();
	let marginLeft = PAD.repeat(margin.left);

	if (options.float === 'center') ***REMOVED***
		const padWidth = Math.max((columns - contentWidth) / 2, 0);
		marginLeft = PAD.repeat(padWidth);
	***REMOVED*** else if (options.float === 'right') ***REMOVED***
		const padWidth = Math.max(columns - contentWidth - margin.right - 2, 0);
		marginLeft = PAD.repeat(padWidth);
	***REMOVED***

	const horizontal = chars.horizontal.repeat(contentWidth);
	const top = colorizeBorder(NL.repeat(margin.top) + marginLeft + chars.topLeft + horizontal + chars.topRight);
	const bottom = colorizeBorder(marginLeft + chars.bottomLeft + horizontal + chars.bottomRight + NL.repeat(margin.bottom));
	const side = colorizeBorder(chars.vertical);

	const middle = lines.map(line => ***REMOVED***
		const paddingRight = PAD.repeat(contentWidth - stringWidth(line) - padding.left);
		return marginLeft + side + colorizeContent(paddingLeft + line + paddingRight) + side;
	***REMOVED***).join(NL);

	return top + NL + middle + NL + bottom;
***REMOVED***;

module.exports._borderStyles = cliBoxes;
