// got/source/errors.js
'use strict';
const urlLib = require('url');
const http = require('http');
const PCancelable = require('p-cancelable');
const is = require('@sindresorhus/is');

class GotError extends Error ***REMOVED***
	constructor(message, error, options) ***REMOVED***
		super(message);
		Error.captureStackTrace(this, this.constructor);
		this.name = 'GotError';

		if (!is.undefined(error.code)) ***REMOVED***
			this.code = error.code;
		***REMOVED***

		Object.assign(this, ***REMOVED***
			host: options.host,
			hostname: options.hostname,
			method: options.method,
			path: options.path,
			socketPath: options.socketPath,
			protocol: options.protocol,
			url: options.href,
			gotOptions: options
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports.GotError = GotError;

module.exports.CacheError = class extends GotError ***REMOVED***
	constructor(error, options) ***REMOVED***
		super(error.message, error, options);
		this.name = 'CacheError';
	***REMOVED***
***REMOVED***;

module.exports.RequestError = class extends GotError ***REMOVED***
	constructor(error, options) ***REMOVED***
		super(error.message, error, options);
		this.name = 'RequestError';
	***REMOVED***
***REMOVED***;

module.exports.ReadError = class extends GotError ***REMOVED***
	constructor(error, options) ***REMOVED***
		super(error.message, error, options);
		this.name = 'ReadError';
	***REMOVED***
***REMOVED***;

module.exports.ParseError = class extends GotError ***REMOVED***
	constructor(error, statusCode, options, data) ***REMOVED***
		super(`$***REMOVED***error.message***REMOVED*** in "$***REMOVED***urlLib.format(options)***REMOVED***": \n$***REMOVED***data.slice(0, 77)***REMOVED***...`, error, options);
		this.name = 'ParseError';
		this.statusCode = statusCode;
		this.statusMessage = http.STATUS_CODES[this.statusCode];
	***REMOVED***
***REMOVED***;

module.exports.HTTPError = class extends GotError ***REMOVED***
	constructor(response, options) ***REMOVED***
		const ***REMOVED***statusCode***REMOVED*** = response;
		let ***REMOVED***statusMessage***REMOVED*** = response;

		if (statusMessage) ***REMOVED***
			statusMessage = statusMessage.replace(/\r?\n/g, ' ').trim();
		***REMOVED*** else ***REMOVED***
			statusMessage = http.STATUS_CODES[statusCode];
		***REMOVED***

		super(`Response code $***REMOVED***statusCode***REMOVED*** ($***REMOVED***statusMessage***REMOVED***)`, ***REMOVED******REMOVED***, options);
		this.name = 'HTTPError';
		this.statusCode = statusCode;
		this.statusMessage = statusMessage;
		this.headers = response.headers;
		this.body = response.body;
	***REMOVED***
***REMOVED***;

module.exports.MaxRedirectsError = class extends GotError ***REMOVED***
	constructor(statusCode, redirectUrls, options) ***REMOVED***
		super('Redirected 10 times. Aborting.', ***REMOVED******REMOVED***, options);
		this.name = 'MaxRedirectsError';
		this.statusCode = statusCode;
		this.statusMessage = http.STATUS_CODES[this.statusCode];
		this.redirectUrls = redirectUrls;
	***REMOVED***
***REMOVED***;

module.exports.UnsupportedProtocolError = class extends GotError ***REMOVED***
	constructor(options) ***REMOVED***
		super(`Unsupported protocol "$***REMOVED***options.protocol***REMOVED***"`, ***REMOVED******REMOVED***, options);
		this.name = 'UnsupportedProtocolError';
	***REMOVED***
***REMOVED***;

module.exports.TimeoutError = class extends GotError ***REMOVED***
	constructor(error, options) ***REMOVED***
		super(error.message, ***REMOVED***code: 'ETIMEDOUT'***REMOVED***, options);
		this.name = 'TimeoutError';
		this.event = error.event;
	***REMOVED***
***REMOVED***;

module.exports.CancelError = PCancelable.CancelError;
