// got/source/index.js
'use strict';
const pkg = require('../package.json');
const create = require('./create');

const defaults = ***REMOVED***
	options: ***REMOVED***
		retry: ***REMOVED***
			retries: 2,
			methods: [
				'GET',
				'PUT',
				'HEAD',
				'DELETE',
				'OPTIONS',
				'TRACE'
			],
			statusCodes: [
				408,
				413,
				429,
				500,
				502,
				503,
				504
			],
			errorCodes: [
				'ETIMEDOUT',
				'ECONNRESET',
				'EADDRINUSE',
				'ECONNREFUSED',
				'EPIPE',
				'ENOTFOUND',
				'ENETUNREACH',
				'EAI_AGAIN'
			]
		***REMOVED***,
		headers: ***REMOVED***
			'user-agent': `$***REMOVED***pkg.name***REMOVED***/$***REMOVED***pkg.version***REMOVED*** (https://github.com/sindresorhus/got)`
		***REMOVED***,
		hooks: ***REMOVED***
			beforeRequest: [],
			beforeRedirect: [],
			beforeRetry: [],
			afterResponse: []
		***REMOVED***,
		decompress: true,
		throwHttpErrors: true,
		followRedirect: true,
		stream: false,
		form: false,
		json: false,
		cache: false,
		useElectronNet: false
	***REMOVED***,
	mutableDefaults: false
***REMOVED***;

const got = create(defaults);

module.exports = got;
