// chalk/index.js
'use strict';
var escapeStringRegexp = require('escape-string-regexp');
var ansiStyles = require('ansi-styles');
var stripAnsi = require('strip-ansi');
var hasAnsi = require('has-ansi');
var supportsColor = require('supports-color');
var defineProps = Object.defineProperties;
var chalk = module.exports;

function build(_styles) ***REMOVED***
	var builder = function builder() ***REMOVED***
		return applyStyle.apply(builder, arguments);
	***REMOVED***;
	builder._styles = _styles;
	// __proto__ is used because we must return a function, but there is
	// no way to create a function with a different prototype.
	builder.__proto__ = proto;
	return builder;
***REMOVED***

var styles = (function () ***REMOVED***
	var ret = ***REMOVED******REMOVED***;

	ansiStyles.grey = ansiStyles.gray;

	Object.keys(ansiStyles).forEach(function (key) ***REMOVED***
		ansiStyles[key].closeRe = new RegExp(escapeStringRegexp(ansiStyles[key].close), 'g');

		ret[key] = ***REMOVED***
			get: function () ***REMOVED***
				return build(this._styles.concat(key));
			***REMOVED***
		***REMOVED***;
	***REMOVED***);

	return ret;
***REMOVED***)();

var proto = defineProps(function chalk() ***REMOVED******REMOVED***, styles);

function applyStyle() ***REMOVED***
	// support varags, but simply cast to string in case there's only one arg
	var args = arguments;
	var argsLen = args.length;
	var str = argsLen !== 0 && String(arguments[0]);
	if (argsLen > 1) ***REMOVED***
		// don't slice `arguments`, it prevents v8 optimizations
		for (var a = 1; a < argsLen; a++) ***REMOVED***
			str += ' ' + args[a];
		***REMOVED***
	***REMOVED***

	if (!chalk.enabled || !str) ***REMOVED***
		return str;
	***REMOVED***

	/*jshint validthis: true*/
	var nestedStyles = this._styles;

	for (var i = 0; i < nestedStyles.length; i++) ***REMOVED***
		var code = ansiStyles[nestedStyles[i]];
		// Replace any instances already present with a re-opening code
		// otherwise only the part of the string until said closing code
		// will be colored, and the rest will simply be 'plain'.
		str = code.open + str.replace(code.closeRe, code.open) + code.close;
	***REMOVED***

	return str;
***REMOVED***

function init() ***REMOVED***
	var ret = ***REMOVED******REMOVED***;

	Object.keys(styles).forEach(function (name) ***REMOVED***
		ret[name] = ***REMOVED***
			get: function () ***REMOVED***
				return build([name]);
			***REMOVED***
		***REMOVED***;
	***REMOVED***);

	return ret;
***REMOVED***

defineProps(chalk, init());

chalk.styles = ansiStyles;
chalk.hasColor = hasAnsi;
chalk.stripColor = stripAnsi;
chalk.supportsColor = supportsColor;

// detect mode if not set manually
if (chalk.enabled === undefined) ***REMOVED***
	chalk.enabled = chalk.supportsColor;
***REMOVED***
