// ejs/lib/utils.js
/*
 * EJS Embedded JavaScript templates
 * Copyright 2112 Matthew Eernisse (mde@fleegix.org)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

/**
 * Private utility functions
 * @module utils
 * @private
 */

'use strict';

var regExpChars = /[|\\***REMOVED******REMOVED***()[\]^$+*?.]/g;

/**
 * Escape characters reserved in regular expressions.
 *
 * If `string` is `undefined` or `null`, the empty string is returned.
 *
 * @param ***REMOVED***String***REMOVED*** string Input string
 * @return ***REMOVED***String***REMOVED*** Escaped string
 * @static
 * @private
 */
exports.escapeRegExpChars = function (string) ***REMOVED***
  // istanbul ignore if
  if (!string) ***REMOVED***
    return '';
  ***REMOVED***
  return String(string).replace(regExpChars, '\\$&');
***REMOVED***;

var _ENCODE_HTML_RULES = ***REMOVED***
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&#34;',
  "'": '&#39;'
***REMOVED***;
var _MATCH_HTML = /[&<>'"]/g;

function encode_char(c) ***REMOVED***
  return _ENCODE_HTML_RULES[c] || c;
***REMOVED***

/**
 * Stringified version of constants used by ***REMOVED***@link module:utils.escapeXML***REMOVED***.
 *
 * It is used in the process of generating ***REMOVED***@link ClientFunction***REMOVED***s.
 *
 * @readonly
 * @type ***REMOVED***String***REMOVED***
 */

var escapeFuncStr =
  'var _ENCODE_HTML_RULES = ***REMOVED***\n'
+ '      "&": "&amp;"\n'
+ '    , "<": "&lt;"\n'
+ '    , ">": "&gt;"\n'
+ '    , \'"\': "&#34;"\n'
+ '    , "\'": "&#39;"\n'
+ '    ***REMOVED***\n'
+ '  , _MATCH_HTML = /[&<>\'"]/g;\n'
+ 'function encode_char(c) ***REMOVED***\n'
+ '  return _ENCODE_HTML_RULES[c] || c;\n'
+ '***REMOVED***;\n';

/**
 * Escape characters reserved in XML.
 *
 * If `markup` is `undefined` or `null`, the empty string is returned.
 *
 * @implements ***REMOVED***EscapeCallback***REMOVED***
 * @param ***REMOVED***String***REMOVED*** markup Input string
 * @return ***REMOVED***String***REMOVED*** Escaped string
 * @static
 * @private
 */

exports.escapeXML = function (markup) ***REMOVED***
  return markup == undefined
    ? ''
    : String(markup)
      .replace(_MATCH_HTML, encode_char);
***REMOVED***;
exports.escapeXML.toString = function () ***REMOVED***
  return Function.prototype.toString.call(this) + ';\n' + escapeFuncStr;
***REMOVED***;

/**
 * Naive copy of properties from one object to another.
 * Does not recurse into non-scalar properties
 * Does not check to see if the property has a value before copying
 *
 * @param  ***REMOVED***Object***REMOVED*** to   Destination object
 * @param  ***REMOVED***Object***REMOVED*** from Source object
 * @return ***REMOVED***Object***REMOVED***      Destination object
 * @static
 * @private
 */
exports.shallowCopy = function (to, from) ***REMOVED***
  from = from || ***REMOVED******REMOVED***;
  for (var p in from) ***REMOVED***
    to[p] = from[p];
  ***REMOVED***
  return to;
***REMOVED***;

/**
 * Naive copy of a list of key names, from one object to another.
 * Only copies property if it is actually defined
 * Does not recurse into non-scalar properties
 *
 * @param  ***REMOVED***Object***REMOVED*** to   Destination object
 * @param  ***REMOVED***Object***REMOVED*** from Source object
 * @param  ***REMOVED***Array***REMOVED*** list List of properties to copy
 * @return ***REMOVED***Object***REMOVED***      Destination object
 * @static
 * @private
 */
exports.shallowCopyFromList = function (to, from, list) ***REMOVED***
  for (var i = 0; i < list.length; i++) ***REMOVED***
    var p = list[i];
    if (typeof from[p] != 'undefined') ***REMOVED***
      to[p] = from[p];
    ***REMOVED***
  ***REMOVED***
  return to;
***REMOVED***;

/**
 * Simple in-process cache implementation. Does not implement limits of any
 * sort.
 *
 * @implements Cache
 * @static
 * @private
 */
exports.cache = ***REMOVED***
  _data: ***REMOVED******REMOVED***,
  set: function (key, val) ***REMOVED***
    this._data[key] = val;
  ***REMOVED***,
  get: function (key) ***REMOVED***
    return this._data[key];
  ***REMOVED***,
  remove: function (key) ***REMOVED***
    delete this._data[key];
  ***REMOVED***,
  reset: function () ***REMOVED***
    this._data = ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED***;
