// rx/dist/rx.coincidence.js
// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved. See License.txt in the project root for license information.

;(function (factory) ***REMOVED***
    var objectTypes = ***REMOVED***
        'boolean': false,
        'function': true,
        'object': true,
        'number': false,
        'string': false,
        'undefined': false
    ***REMOVED***;

    var root = (objectTypes[typeof window] && window) || this,
        freeExports = objectTypes[typeof exports] && exports && !exports.nodeType && exports,
        freeModule = objectTypes[typeof module] && module && !module.nodeType && module,
        moduleExports = freeModule && freeModule.exports === freeExports && freeExports,
        freeGlobal = objectTypes[typeof global] && global;

    if (freeGlobal && (freeGlobal.global === freeGlobal || freeGlobal.window === freeGlobal)) ***REMOVED***
        root = freeGlobal;
    ***REMOVED***

    // Because of build optimizers
    if (typeof define === 'function' && define.amd) ***REMOVED***
        define(['rx'], function (Rx, exports) ***REMOVED***
            return factory(root, exports, Rx);
        ***REMOVED***);
    ***REMOVED*** else if (typeof module === 'object' && module && module.exports === freeExports) ***REMOVED***
        module.exports = factory(root, module.exports, require('./rx'));
    ***REMOVED*** else ***REMOVED***
        root.Rx = factory(root, ***REMOVED******REMOVED***, root.Rx);
    ***REMOVED***
***REMOVED***.call(this, function (root, exp, Rx, undefined) ***REMOVED***

  var Observable = Rx.Observable,
    CompositeDisposable = Rx.CompositeDisposable,
    RefCountDisposable = Rx.RefCountDisposable,
    SingleAssignmentDisposable = Rx.SingleAssignmentDisposable,
    SerialDisposable = Rx.SerialDisposable,
    Subject = Rx.Subject,
    observableProto = Observable.prototype,
    observableEmpty = Observable.empty,
    observableNever = Observable.never,
    AnonymousObservable = Rx.AnonymousObservable,
    observerCreate = Rx.Observer.create,
    addRef = Rx.internals.addRef,
    defaultComparer = Rx.internals.isEqual,
    noop = Rx.helpers.noop,
    identity = Rx.helpers.identity,
    isPromise = Rx.helpers.isPromise,
    observableFromPromise = Observable.fromPromise;

  var Dictionary = (function () ***REMOVED***

    var primes = [1, 3, 7, 13, 31, 61, 127, 251, 509, 1021, 2039, 4093, 8191, 16381, 32749, 65521, 131071, 262139, 524287, 1048573, 2097143, 4194301, 8388593, 16777213, 33554393, 67108859, 134217689, 268435399, 536870909, 1073741789, 2147483647],
      noSuchkey = "no such key",
      duplicatekey = "duplicate key";

    function isPrime(candidate) ***REMOVED***
      if ((candidate & 1) === 0) ***REMOVED*** return candidate === 2; ***REMOVED***
      var num1 = Math.sqrt(candidate),
        num2 = 3;
      while (num2 <= num1) ***REMOVED***
        if (candidate % num2 === 0) ***REMOVED*** return false; ***REMOVED***
        num2 += 2;
      ***REMOVED***
      return true;
    ***REMOVED***

    function getPrime(min) ***REMOVED***
      var index, num, candidate;
      for (index = 0; index < primes.length; ++index) ***REMOVED***
        num = primes[index];
        if (num >= min) ***REMOVED*** return num; ***REMOVED***
      ***REMOVED***
      candidate = min | 1;
      while (candidate < primes[primes.length - 1]) ***REMOVED***
        if (isPrime(candidate)) ***REMOVED*** return candidate; ***REMOVED***
        candidate += 2;
      ***REMOVED***
      return min;
    ***REMOVED***

    function stringHashFn(str) ***REMOVED***
      var hash = 757602046;
      if (!str.length) ***REMOVED*** return hash; ***REMOVED***
      for (var i = 0, len = str.length; i < len; i++) ***REMOVED***
        var character = str.charCodeAt(i);
        hash = ((hash << 5) - hash) + character;
        hash = hash & hash;
      ***REMOVED***
      return hash;
    ***REMOVED***

    function numberHashFn(key) ***REMOVED***
      var c2 = 0x27d4eb2d;
      key = (key ^ 61) ^ (key >>> 16);
      key = key + (key << 3);
      key = key ^ (key >>> 4);
      key = key * c2;
      key = key ^ (key >>> 15);
      return key;
    ***REMOVED***

    var getHashCode = (function () ***REMOVED***
      var uniqueIdCounter = 0;

      return function (obj) ***REMOVED***
        if (obj == null) ***REMOVED*** throw new Error(noSuchkey); ***REMOVED***

        // Check for built-ins before tacking on our own for any object
        if (typeof obj === 'string') ***REMOVED*** return stringHashFn(obj); ***REMOVED***
        if (typeof obj === 'number') ***REMOVED*** return numberHashFn(obj); ***REMOVED***
        if (typeof obj === 'boolean') ***REMOVED*** return obj === true ? 1 : 0; ***REMOVED***
        if (obj instanceof Date) ***REMOVED*** return numberHashFn(obj.valueOf()); ***REMOVED***
        if (obj instanceof RegExp) ***REMOVED*** return stringHashFn(obj.toString()); ***REMOVED***
        if (typeof obj.valueOf === 'function') ***REMOVED***
          // Hack check for valueOf
          var valueOf = obj.valueOf();
          if (typeof valueOf === 'number') ***REMOVED*** return numberHashFn(valueOf); ***REMOVED***
          if (typeof obj === 'string') ***REMOVED*** return stringHashFn(valueOf); ***REMOVED***
        ***REMOVED***
        if (obj.hashCode) ***REMOVED*** return obj.hashCode(); ***REMOVED***

        var id = 17 * uniqueIdCounter++;
        obj.hashCode = function () ***REMOVED*** return id; ***REMOVED***;
        return id;
      ***REMOVED***;
    ***REMOVED***());

    function newEntry() ***REMOVED***
      return ***REMOVED*** key: null, value: null, next: 0, hashCode: 0 ***REMOVED***;
    ***REMOVED***

    function Dictionary(capacity, comparer) ***REMOVED***
      if (capacity < 0) ***REMOVED*** throw new Error('out of range'); ***REMOVED***
      if (capacity > 0) ***REMOVED*** this._initialize(capacity); ***REMOVED***

      this.comparer = comparer || defaultComparer;
      this.freeCount = 0;
      this.size = 0;
      this.freeList = -1;
    ***REMOVED***

    var dictionaryProto = Dictionary.prototype;

    dictionaryProto._initialize = function (capacity) ***REMOVED***
      var prime = getPrime(capacity), i;
      this.buckets = new Array(prime);
      this.entries = new Array(prime);
      for (i = 0; i < prime; i++) ***REMOVED***
        this.buckets[i] = -1;
        this.entries[i] = newEntry();
      ***REMOVED***
      this.freeList = -1;
    ***REMOVED***;

    dictionaryProto.add = function (key, value) ***REMOVED***
      this._insert(key, value, true);
    ***REMOVED***;

    dictionaryProto._insert = function (key, value, add) ***REMOVED***
      if (!this.buckets) ***REMOVED*** this._initialize(0); ***REMOVED***
      var index3,
        num = getHashCode(key) & 2147483647,
        index1 = num % this.buckets.length;
      for (var index2 = this.buckets[index1]; index2 >= 0; index2 = this.entries[index2].next) ***REMOVED***
        if (this.entries[index2].hashCode === num && this.comparer(this.entries[index2].key, key)) ***REMOVED***
          if (add) ***REMOVED*** throw new Error(duplicatekey); ***REMOVED***
          this.entries[index2].value = value;
          return;
        ***REMOVED***
      ***REMOVED***
      if (this.freeCount > 0) ***REMOVED***
        index3 = this.freeList;
        this.freeList = this.entries[index3].next;
        --this.freeCount;
      ***REMOVED*** else ***REMOVED***
        if (this.size === this.entries.length) ***REMOVED***
          this._resize();
          index1 = num % this.buckets.length;
        ***REMOVED***
        index3 = this.size;
        ++this.size;
      ***REMOVED***
      this.entries[index3].hashCode = num;
      this.entries[index3].next = this.buckets[index1];
      this.entries[index3].key = key;
      this.entries[index3].value = value;
      this.buckets[index1] = index3;
    ***REMOVED***;

    dictionaryProto._resize = function () ***REMOVED***
      var prime = getPrime(this.size * 2),
        numArray = new Array(prime);
      for (index = 0; index < numArray.length; ++index) ***REMOVED***  numArray[index] = -1; ***REMOVED***
      var entryArray = new Array(prime);
      for (index = 0; index < this.size; ++index) ***REMOVED*** entryArray[index] = this.entries[index]; ***REMOVED***
      for (var index = this.size; index < prime; ++index) ***REMOVED*** entryArray[index] = newEntry(); ***REMOVED***
      for (var index1 = 0; index1 < this.size; ++index1) ***REMOVED***
        var index2 = entryArray[index1].hashCode % prime;
        entryArray[index1].next = numArray[index2];
        numArray[index2] = index1;
      ***REMOVED***
      this.buckets = numArray;
      this.entries = entryArray;
    ***REMOVED***;

    dictionaryProto.remove = function (key) ***REMOVED***
      if (this.buckets) ***REMOVED***
        var num = getHashCode(key) & 2147483647,
          index1 = num % this.buckets.length,
          index2 = -1;
        for (var index3 = this.buckets[index1]; index3 >= 0; index3 = this.entries[index3].next) ***REMOVED***
          if (this.entries[index3].hashCode === num && this.comparer(this.entries[index3].key, key)) ***REMOVED***
            if (index2 < 0) ***REMOVED***
              this.buckets[index1] = this.entries[index3].next;
            ***REMOVED*** else ***REMOVED***
              this.entries[index2].next = this.entries[index3].next;
            ***REMOVED***
            this.entries[index3].hashCode = -1;
            this.entries[index3].next = this.freeList;
            this.entries[index3].key = null;
            this.entries[index3].value = null;
            this.freeList = index3;
            ++this.freeCount;
            return true;
          ***REMOVED*** else ***REMOVED***
            index2 = index3;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return false;
    ***REMOVED***;

    dictionaryProto.clear = function () ***REMOVED***
      var index, len;
      if (this.size <= 0) ***REMOVED*** return; ***REMOVED***
      for (index = 0, len = this.buckets.length; index < len; ++index) ***REMOVED***
        this.buckets[index] = -1;
      ***REMOVED***
      for (index = 0; index < this.size; ++index) ***REMOVED***
        this.entries[index] = newEntry();
      ***REMOVED***
      this.freeList = -1;
      this.size = 0;
    ***REMOVED***;

    dictionaryProto._findEntry = function (key) ***REMOVED***
      if (this.buckets) ***REMOVED***
        var num = getHashCode(key) & 2147483647;
        for (var index = this.buckets[num % this.buckets.length]; index >= 0; index = this.entries[index].next) ***REMOVED***
          if (this.entries[index].hashCode === num && this.comparer(this.entries[index].key, key)) ***REMOVED***
            return index;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return -1;
    ***REMOVED***;

    dictionaryProto.count = function () ***REMOVED***
      return this.size - this.freeCount;
    ***REMOVED***;

    dictionaryProto.tryGetValue = function (key) ***REMOVED***
      var entry = this._findEntry(key);
      return entry >= 0 ?
        this.entries[entry].value :
        undefined;
    ***REMOVED***;

    dictionaryProto.getValues = function () ***REMOVED***
      var index = 0, results = [];
      if (this.entries) ***REMOVED***
        for (var index1 = 0; index1 < this.size; index1++) ***REMOVED***
          if (this.entries[index1].hashCode >= 0) ***REMOVED***
            results[index++] = this.entries[index1].value;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return results;
    ***REMOVED***;

    dictionaryProto.get = function (key) ***REMOVED***
      var entry = this._findEntry(key);
      if (entry >= 0) ***REMOVED*** return this.entries[entry].value; ***REMOVED***
      throw new Error(noSuchkey);
    ***REMOVED***;

    dictionaryProto.set = function (key, value) ***REMOVED***
      this._insert(key, value, false);
    ***REMOVED***;

    dictionaryProto.containskey = function (key) ***REMOVED***
      return this._findEntry(key) >= 0;
    ***REMOVED***;

    return Dictionary;
  ***REMOVED***());

  /**
   *  Correlates the elements of two sequences based on overlapping durations.
   *
   *  @param ***REMOVED***Observable***REMOVED*** right The right observable sequence to join elements for.
   *  @param ***REMOVED***Function***REMOVED*** leftDurationSelector A function to select the duration (expressed as an observable sequence) of each element of the left observable sequence, used to determine overlap.
   *  @param ***REMOVED***Function***REMOVED*** rightDurationSelector A function to select the duration (expressed as an observable sequence) of each element of the right observable sequence, used to determine overlap.
   *  @param ***REMOVED***Function***REMOVED*** resultSelector A function invoked to compute a result element for any two overlapping elements of the left and right observable sequences. The parameters passed to the function correspond with the elements from the left and right source sequences for which overlap occurs.
   *  @returns ***REMOVED***Observable***REMOVED*** An observable sequence that contains result elements computed from source elements that have an overlapping duration.
   */
  observableProto.join = function (right, leftDurationSelector, rightDurationSelector, resultSelector) ***REMOVED***
    var left = this;
    return new AnonymousObservable(function (observer) ***REMOVED***
      var group = new CompositeDisposable();
      var leftDone = false, rightDone = false;
      var leftId = 0, rightId = 0;
      var leftMap = new Dictionary(), rightMap = new Dictionary();

      group.add(left.subscribe(
        function (value) ***REMOVED***
          var id = leftId++;
          var md = new SingleAssignmentDisposable();

          leftMap.add(id, value);
          group.add(md);

          var expire = function () ***REMOVED***
            leftMap.remove(id) && leftMap.count() === 0 && leftDone && observer.onCompleted();
            group.remove(md);
          ***REMOVED***;

          var duration;
          try ***REMOVED***
            duration = leftDurationSelector(value);
          ***REMOVED*** catch (e) ***REMOVED***
            observer.onError(e);
            return;
          ***REMOVED***

          md.setDisposable(duration.take(1).subscribe(noop, observer.onError.bind(observer), expire));

          rightMap.getValues().forEach(function (v) ***REMOVED***
            var result;
            try ***REMOVED***
              result = resultSelector(value, v);
            ***REMOVED*** catch (exn) ***REMOVED***
              observer.onError(exn);
              return;
            ***REMOVED***

            observer.onNext(result);
          ***REMOVED***);
        ***REMOVED***,
        observer.onError.bind(observer),
        function () ***REMOVED***
          leftDone = true;
          (rightDone || leftMap.count() === 0) && observer.onCompleted();
        ***REMOVED***)
      );

      group.add(right.subscribe(
        function (value) ***REMOVED***
          var id = rightId++;
          var md = new SingleAssignmentDisposable();

          rightMap.add(id, value);
          group.add(md);

          var expire = function () ***REMOVED***
            rightMap.remove(id) && rightMap.count() === 0 && rightDone && observer.onCompleted();
            group.remove(md);
          ***REMOVED***;

          var duration;
          try ***REMOVED***
            duration = rightDurationSelector(value);
          ***REMOVED*** catch (e) ***REMOVED***
            observer.onError(e);
            return;
          ***REMOVED***

          md.setDisposable(duration.take(1).subscribe(noop, observer.onError.bind(observer), expire));

          leftMap.getValues().forEach(function (v) ***REMOVED***
            var result;
            try ***REMOVED***
              result = resultSelector(v, value);
            ***REMOVED*** catch (exn) ***REMOVED***
              observer.onError(exn);
              return;
            ***REMOVED***

            observer.onNext(result);
          ***REMOVED***);
        ***REMOVED***,
        observer.onError.bind(observer),
        function () ***REMOVED***
          rightDone = true;
          (leftDone || rightMap.count() === 0) && observer.onCompleted();
        ***REMOVED***)
      );
      return group;
    ***REMOVED***, left);
  ***REMOVED***;

  /**
   *  Correlates the elements of two sequences based on overlapping durations, and groups the results.
   *
   *  @param ***REMOVED***Observable***REMOVED*** right The right observable sequence to join elements for.
   *  @param ***REMOVED***Function***REMOVED*** leftDurationSelector A function to select the duration (expressed as an observable sequence) of each element of the left observable sequence, used to determine overlap.
   *  @param ***REMOVED***Function***REMOVED*** rightDurationSelector A function to select the duration (expressed as an observable sequence) of each element of the right observable sequence, used to determine overlap.
   *  @param ***REMOVED***Function***REMOVED*** resultSelector A function invoked to compute a result element for any element of the left sequence with overlapping elements from the right observable sequence. The first parameter passed to the function is an element of the left sequence. The second parameter passed to the function is an observable sequence with elements from the right sequence that overlap with the left sequence's element.
   *  @returns ***REMOVED***Observable***REMOVED*** An observable sequence that contains result elements computed from source elements that have an overlapping duration.
   */
  observableProto.groupJoin = function (right, leftDurationSelector, rightDurationSelector, resultSelector) ***REMOVED***
    var left = this;
    return new AnonymousObservable(function (observer) ***REMOVED***
      var group = new CompositeDisposable();
      var r = new RefCountDisposable(group);
      var leftMap = new Dictionary(), rightMap = new Dictionary();
      var leftId = 0, rightId = 0;

      function handleError(e) ***REMOVED*** return function (v) ***REMOVED*** v.onError(e); ***REMOVED***; ***REMOVED***;

      group.add(left.subscribe(
        function (value) ***REMOVED***
          var s = new Subject();
          var id = leftId++;
          leftMap.add(id, s);

          var result;
          try ***REMOVED***
            result = resultSelector(value, addRef(s, r));
          ***REMOVED*** catch (e) ***REMOVED***
            leftMap.getValues().forEach(handleError(e));
            observer.onError(e);
            return;
          ***REMOVED***
          observer.onNext(result);

          rightMap.getValues().forEach(function (v) ***REMOVED*** s.onNext(v); ***REMOVED***);

          var md = new SingleAssignmentDisposable();
          group.add(md);

          var expire = function () ***REMOVED***
            leftMap.remove(id) && s.onCompleted();
            group.remove(md);
          ***REMOVED***;

          var duration;
          try ***REMOVED***
            duration = leftDurationSelector(value);
          ***REMOVED*** catch (e) ***REMOVED***
            leftMap.getValues().forEach(handleError(e));
            observer.onError(e);
            return;
          ***REMOVED***

          md.setDisposable(duration.take(1).subscribe(
            noop,
            function (e) ***REMOVED***
              leftMap.getValues().forEach(handleError(e));
              observer.onError(e);
            ***REMOVED***,
            expire)
          );
        ***REMOVED***,
        function (e) ***REMOVED***
          leftMap.getValues().forEach(handleError(e));
          observer.onError(e);
        ***REMOVED***,
        observer.onCompleted.bind(observer))
      );

      group.add(right.subscribe(
        function (value) ***REMOVED***
          var id = rightId++;
          rightMap.add(id, value);

          var md = new SingleAssignmentDisposable();
          group.add(md);

          var expire = function () ***REMOVED***
            rightMap.remove(id);
            group.remove(md);
          ***REMOVED***;

          var duration;
          try ***REMOVED***
            duration = rightDurationSelector(value);
          ***REMOVED*** catch (e) ***REMOVED***
            leftMap.getValues().forEach(handleError(e));
            observer.onError(e);
            return;
          ***REMOVED***
          md.setDisposable(duration.take(1).subscribe(
            noop,
            function (e) ***REMOVED***
              leftMap.getValues().forEach(handleError(e));
              observer.onError(e);
            ***REMOVED***,
            expire)
          );

          leftMap.getValues().forEach(function (v) ***REMOVED*** v.onNext(value); ***REMOVED***);
        ***REMOVED***,
        function (e) ***REMOVED***
          leftMap.getValues().forEach(handleError(e));
          observer.onError(e);
        ***REMOVED***)
      );

      return r;
    ***REMOVED***, left);
  ***REMOVED***;

    /**
     *  Projects each element of an observable sequence into zero or more buffers.
     *
     *  @param ***REMOVED***Mixed***REMOVED*** bufferOpeningsOrClosingSelector Observable sequence whose elements denote the creation of new windows, or, a function invoked to define the boundaries of the produced windows (a new window is started when the previous one is closed, resulting in non-overlapping windows).
     *  @param ***REMOVED***Function***REMOVED*** [bufferClosingSelector] A function invoked to define the closing of each produced window. If a closing selector function is specified for the first parameter, this parameter is ignored.
     *  @returns ***REMOVED***Observable***REMOVED*** An observable sequence of windows.
     */
    observableProto.buffer = function (bufferOpeningsOrClosingSelector, bufferClosingSelector) ***REMOVED***
        return this.window.apply(this, arguments).selectMany(function (x) ***REMOVED*** return x.toArray(); ***REMOVED***);
    ***REMOVED***;

  /**
   *  Projects each element of an observable sequence into zero or more windows.
   *
   *  @param ***REMOVED***Mixed***REMOVED*** windowOpeningsOrClosingSelector Observable sequence whose elements denote the creation of new windows, or, a function invoked to define the boundaries of the produced windows (a new window is started when the previous one is closed, resulting in non-overlapping windows).
   *  @param ***REMOVED***Function***REMOVED*** [windowClosingSelector] A function invoked to define the closing of each produced window. If a closing selector function is specified for the first parameter, this parameter is ignored.
   *  @returns ***REMOVED***Observable***REMOVED*** An observable sequence of windows.
   */
  observableProto.window = function (windowOpeningsOrClosingSelector, windowClosingSelector) ***REMOVED***
    if (arguments.length === 1 && typeof arguments[0] !== 'function') ***REMOVED***
      return observableWindowWithBoundaries.call(this, windowOpeningsOrClosingSelector);
    ***REMOVED***
    return typeof windowOpeningsOrClosingSelector === 'function' ?
      observableWindowWithClosingSelector.call(this, windowOpeningsOrClosingSelector) :
      observableWindowWithOpenings.call(this, windowOpeningsOrClosingSelector, windowClosingSelector);
  ***REMOVED***;

  function observableWindowWithOpenings(windowOpenings, windowClosingSelector) ***REMOVED***
    return windowOpenings.groupJoin(this, windowClosingSelector, observableEmpty, function (_, win) ***REMOVED***
      return win;
    ***REMOVED***);
  ***REMOVED***

  function observableWindowWithBoundaries(windowBoundaries) ***REMOVED***
    var source = this;
    return new AnonymousObservable(function (observer) ***REMOVED***
      var win = new Subject(),
        d = new CompositeDisposable(),
        r = new RefCountDisposable(d);

      observer.onNext(addRef(win, r));

      d.add(source.subscribe(function (x) ***REMOVED***
        win.onNext(x);
      ***REMOVED***, function (err) ***REMOVED***
        win.onError(err);
        observer.onError(err);
      ***REMOVED***, function () ***REMOVED***
        win.onCompleted();
        observer.onCompleted();
      ***REMOVED***));

      isPromise(windowBoundaries) && (windowBoundaries = observableFromPromise(windowBoundaries));

      d.add(windowBoundaries.subscribe(function (w) ***REMOVED***
        win.onCompleted();
        win = new Subject();
        observer.onNext(addRef(win, r));
      ***REMOVED***, function (err) ***REMOVED***
        win.onError(err);
        observer.onError(err);
      ***REMOVED***, function () ***REMOVED***
        win.onCompleted();
        observer.onCompleted();
      ***REMOVED***));

      return r;
    ***REMOVED***, source);
  ***REMOVED***

  function observableWindowWithClosingSelector(windowClosingSelector) ***REMOVED***
    var source = this;
    return new AnonymousObservable(function (observer) ***REMOVED***
      var m = new SerialDisposable(),
        d = new CompositeDisposable(m),
        r = new RefCountDisposable(d),
        win = new Subject();
      observer.onNext(addRef(win, r));
      d.add(source.subscribe(function (x) ***REMOVED***
          win.onNext(x);
      ***REMOVED***, function (err) ***REMOVED***
          win.onError(err);
          observer.onError(err);
      ***REMOVED***, function () ***REMOVED***
          win.onCompleted();
          observer.onCompleted();
      ***REMOVED***));

      function createWindowClose () ***REMOVED***
        var windowClose;
        try ***REMOVED***
          windowClose = windowClosingSelector();
        ***REMOVED*** catch (e) ***REMOVED***
          observer.onError(e);
          return;
        ***REMOVED***

        isPromise(windowClose) && (windowClose = observableFromPromise(windowClose));

        var m1 = new SingleAssignmentDisposable();
        m.setDisposable(m1);
        m1.setDisposable(windowClose.take(1).subscribe(noop, function (err) ***REMOVED***
          win.onError(err);
          observer.onError(err);
        ***REMOVED***, function () ***REMOVED***
          win.onCompleted();
          win = new Subject();
          observer.onNext(addRef(win, r));
          createWindowClose();
        ***REMOVED***));
      ***REMOVED***

      createWindowClose();
      return r;
    ***REMOVED***, source);
  ***REMOVED***

  /**
   * Returns a new observable that triggers on the second and subsequent triggerings of the input observable.
   * The Nth triggering of the input observable passes the arguments from the N-1th and Nth triggering as a pair.
   * The argument passed to the N-1th triggering is held in hidden internal state until the Nth triggering occurs.
   * @returns ***REMOVED***Observable***REMOVED*** An observable that triggers on successive pairs of observations from the input observable as an array.
   */
  observableProto.pairwise = function () ***REMOVED***
    var source = this;
    return new AnonymousObservable(function (observer) ***REMOVED***
      var previous, hasPrevious = false;
      return source.subscribe(
        function (x) ***REMOVED***
          if (hasPrevious) ***REMOVED***
            observer.onNext([previous, x]);
          ***REMOVED*** else ***REMOVED***
            hasPrevious = true;
          ***REMOVED***
          previous = x;
        ***REMOVED***,
        observer.onError.bind(observer),
        observer.onCompleted.bind(observer));
    ***REMOVED***, source);
  ***REMOVED***;

  /**
   * Returns two observables which partition the observations of the source by the given function.
   * The first will trigger observations for those values for which the predicate returns true.
   * The second will trigger observations for those values where the predicate returns false.
   * The predicate is executed once for each subscribed observer.
   * Both also propagate all error observations arising from the source and each completes
   * when the source completes.
   * @param ***REMOVED***Function***REMOVED*** predicate
   *    The function to determine which output Observable will trigger a particular observation.
   * @returns ***REMOVED***Array***REMOVED***
   *    An array of observables. The first triggers when the predicate returns true,
   *    and the second triggers when the predicate returns false.
  */
  observableProto.partition = function(predicate, thisArg) ***REMOVED***
    return [
      this.filter(predicate, thisArg),
      this.filter(function (x, i, o) ***REMOVED*** return !predicate.call(thisArg, x, i, o); ***REMOVED***)
    ];
  ***REMOVED***;

  /**
   *  Groups the elements of an observable sequence according to a specified key selector function and comparer and selects the resulting elements by using a specified function.
   *
   * @example
   *  var res = observable.groupBy(function (x) ***REMOVED*** return x.id; ***REMOVED***);
   *  2 - observable.groupBy(function (x) ***REMOVED*** return x.id; ***REMOVED***), function (x) ***REMOVED*** return x.name; ***REMOVED***);
   *  3 - observable.groupBy(function (x) ***REMOVED*** return x.id; ***REMOVED***), function (x) ***REMOVED*** return x.name; ***REMOVED***, function (x) ***REMOVED*** return x.toString(); ***REMOVED***);
   * @param ***REMOVED***Function***REMOVED*** keySelector A function to extract the key for each element.
   * @param ***REMOVED***Function***REMOVED*** [elementSelector]  A function to map each source element to an element in an observable group.
   * @param ***REMOVED***Function***REMOVED*** [comparer] Used to determine whether the objects are equal.
   * @returns ***REMOVED***Observable***REMOVED*** A sequence of observable groups, each of which corresponds to a unique key value, containing all elements that share that same key value.
   */
  observableProto.groupBy = function (keySelector, elementSelector, comparer) ***REMOVED***
    return this.groupByUntil(keySelector, elementSelector, observableNever, comparer);
  ***REMOVED***;

    /**
     *  Groups the elements of an observable sequence according to a specified key selector function.
     *  A duration selector function is used to control the lifetime of groups. When a group expires, it receives an OnCompleted notification. When a new element with the same
     *  key value as a reclaimed group occurs, the group will be reborn with a new lifetime request.
     *
     * @example
     *  var res = observable.groupByUntil(function (x) ***REMOVED*** return x.id; ***REMOVED***, null,  function () ***REMOVED*** return Rx.Observable.never(); ***REMOVED***);
     *  2 - observable.groupBy(function (x) ***REMOVED*** return x.id; ***REMOVED***), function (x) ***REMOVED*** return x.name; ***REMOVED***,  function () ***REMOVED*** return Rx.Observable.never(); ***REMOVED***);
     *  3 - observable.groupBy(function (x) ***REMOVED*** return x.id; ***REMOVED***), function (x) ***REMOVED*** return x.name; ***REMOVED***,  function () ***REMOVED*** return Rx.Observable.never(); ***REMOVED***, function (x) ***REMOVED*** return x.toString(); ***REMOVED***);
     * @param ***REMOVED***Function***REMOVED*** keySelector A function to extract the key for each element.
     * @param ***REMOVED***Function***REMOVED*** durationSelector A function to signal the expiration of a group.
     * @param ***REMOVED***Function***REMOVED*** [comparer] Used to compare objects. When not specified, the default comparer is used.
     * @returns ***REMOVED***Observable***REMOVED***
     *  A sequence of observable groups, each of which corresponds to a unique key value, containing all elements that share that same key value.
     *  If a group's lifetime expires, a new group with the same key value can be created once an element with such a key value is encoutered.
     *
     */
    observableProto.groupByUntil = function (keySelector, elementSelector, durationSelector, comparer) ***REMOVED***
      var source = this;
      elementSelector || (elementSelector = identity);
      comparer || (comparer = defaultComparer);
      return new AnonymousObservable(function (observer) ***REMOVED***
        function handleError(e) ***REMOVED*** return function (item) ***REMOVED*** item.onError(e); ***REMOVED***; ***REMOVED***
        var map = new Dictionary(0, comparer),
          groupDisposable = new CompositeDisposable(),
          refCountDisposable = new RefCountDisposable(groupDisposable);

        groupDisposable.add(source.subscribe(function (x) ***REMOVED***
          var key;
          try ***REMOVED***
            key = keySelector(x);
          ***REMOVED*** catch (e) ***REMOVED***
            map.getValues().forEach(handleError(e));
            observer.onError(e);
            return;
          ***REMOVED***

          var fireNewMapEntry = false,
            writer = map.tryGetValue(key);
          if (!writer) ***REMOVED***
            writer = new Subject();
            map.set(key, writer);
            fireNewMapEntry = true;
          ***REMOVED***

          if (fireNewMapEntry) ***REMOVED***
            var group = new GroupedObservable(key, writer, refCountDisposable),
              durationGroup = new GroupedObservable(key, writer);
            try ***REMOVED***
              duration = durationSelector(durationGroup);
            ***REMOVED*** catch (e) ***REMOVED***
              map.getValues().forEach(handleError(e));
              observer.onError(e);
              return;
            ***REMOVED***

            observer.onNext(group);

            var md = new SingleAssignmentDisposable();
            groupDisposable.add(md);

            var expire = function () ***REMOVED***
              map.remove(key) && writer.onCompleted();
              groupDisposable.remove(md);
            ***REMOVED***;

            md.setDisposable(duration.take(1).subscribe(
              noop,
              function (exn) ***REMOVED***
                map.getValues().forEach(handleError(exn));
                observer.onError(exn);
              ***REMOVED***,
              expire)
            );
          ***REMOVED***

          var element;
          try ***REMOVED***
            element = elementSelector(x);
          ***REMOVED*** catch (e) ***REMOVED***
            map.getValues().forEach(handleError(e));
            observer.onError(e);
            return;
          ***REMOVED***

          writer.onNext(element);
      ***REMOVED***, function (ex) ***REMOVED***
        map.getValues().forEach(handleError(ex));
        observer.onError(ex);
      ***REMOVED***, function () ***REMOVED***
        map.getValues().forEach(function (item) ***REMOVED*** item.onCompleted(); ***REMOVED***);
        observer.onCompleted();
      ***REMOVED***));

      return refCountDisposable;
    ***REMOVED***, source);
  ***REMOVED***;

  var GroupedObservable = (function (__super__) ***REMOVED***
    inherits(GroupedObservable, __super__);

    function subscribe(observer) ***REMOVED***
      return this.underlyingObservable.subscribe(observer);
    ***REMOVED***

    function GroupedObservable(key, underlyingObservable, mergedDisposable) ***REMOVED***
      __super__.call(this, subscribe);
      this.key = key;
      this.underlyingObservable = !mergedDisposable ?
        underlyingObservable :
        new AnonymousObservable(function (observer) ***REMOVED***
          return new CompositeDisposable(mergedDisposable.getDisposable(), underlyingObservable.subscribe(observer));
        ***REMOVED***);
    ***REMOVED***

    return GroupedObservable;
  ***REMOVED***(Observable));

    return Rx;
***REMOVED***));
