// date-fns/parse/index.js
var getTimezoneOffsetInMilliseconds = require('../_lib/getTimezoneOffsetInMilliseconds/index.js')
var isDate = require('../is_date/index.js')

var MILLISECONDS_IN_HOUR = 3600000
var MILLISECONDS_IN_MINUTE = 60000
var DEFAULT_ADDITIONAL_DIGITS = 2

var parseTokenDateTimeDelimeter = /[T ]/
var parseTokenPlainTime = /:/

// year tokens
var parseTokenYY = /^(\d***REMOVED***2***REMOVED***)$/
var parseTokensYYY = [
  /^([+-]\d***REMOVED***2***REMOVED***)$/, // 0 additional digits
  /^([+-]\d***REMOVED***3***REMOVED***)$/, // 1 additional digit
  /^([+-]\d***REMOVED***4***REMOVED***)$/ // 2 additional digits
]

var parseTokenYYYY = /^(\d***REMOVED***4***REMOVED***)/
var parseTokensYYYYY = [
  /^([+-]\d***REMOVED***4***REMOVED***)/, // 0 additional digits
  /^([+-]\d***REMOVED***5***REMOVED***)/, // 1 additional digit
  /^([+-]\d***REMOVED***6***REMOVED***)/ // 2 additional digits
]

// date tokens
var parseTokenMM = /^-(\d***REMOVED***2***REMOVED***)$/
var parseTokenDDD = /^-?(\d***REMOVED***3***REMOVED***)$/
var parseTokenMMDD = /^-?(\d***REMOVED***2***REMOVED***)-?(\d***REMOVED***2***REMOVED***)$/
var parseTokenWww = /^-?W(\d***REMOVED***2***REMOVED***)$/
var parseTokenWwwD = /^-?W(\d***REMOVED***2***REMOVED***)-?(\d***REMOVED***1***REMOVED***)$/

// time tokens
var parseTokenHH = /^(\d***REMOVED***2***REMOVED***([.,]\d*)?)$/
var parseTokenHHMM = /^(\d***REMOVED***2***REMOVED***):?(\d***REMOVED***2***REMOVED***([.,]\d*)?)$/
var parseTokenHHMMSS = /^(\d***REMOVED***2***REMOVED***):?(\d***REMOVED***2***REMOVED***):?(\d***REMOVED***2***REMOVED***([.,]\d*)?)$/

// timezone tokens
var parseTokenTimezone = /([Z+-].*)$/
var parseTokenTimezoneZ = /^(Z)$/
var parseTokenTimezoneHH = /^([+-])(\d***REMOVED***2***REMOVED***)$/
var parseTokenTimezoneHHMM = /^([+-])(\d***REMOVED***2***REMOVED***):?(\d***REMOVED***2***REMOVED***)$/

/**
 * @category Common Helpers
 * @summary Convert the given argument to an instance of Date.
 *
 * @description
 * Convert the given argument to an instance of Date.
 *
 * If the argument is an instance of Date, the function returns its clone.
 *
 * If the argument is a number, it is treated as a timestamp.
 *
 * If an argument is a string, the function tries to parse it.
 * Function accepts complete ISO 8601 formats as well as partial implementations.
 * ISO 8601: http://en.wikipedia.org/wiki/ISO_8601
 *
 * If all above fails, the function passes the given argument to Date constructor.
 *
 * @param ***REMOVED***Date|String|Number***REMOVED*** argument - the value to convert
 * @param ***REMOVED***Object***REMOVED*** [options] - the object with options
 * @param ***REMOVED***0 | 1 | 2***REMOVED*** [options.additionalDigits=2] - the additional number of digits in the extended year format
 * @returns ***REMOVED***Date***REMOVED*** the parsed date in the local time zone
 *
 * @example
 * // Convert string '2014-02-11T11:30:30' to date:
 * var result = parse('2014-02-11T11:30:30')
 * //=> Tue Feb 11 2014 11:30:30
 *
 * @example
 * // Parse string '+02014101',
 * // if the additional number of digits in the extended year format is 1:
 * var result = parse('+02014101', ***REMOVED***additionalDigits: 1***REMOVED***)
 * //=> Fri Apr 11 2014 00:00:00
 */
function parse (argument, dirtyOptions) ***REMOVED***
  if (isDate(argument)) ***REMOVED***
    // Prevent the date to lose the milliseconds when passed to new Date() in IE10
    return new Date(argument.getTime())
  ***REMOVED*** else if (typeof argument !== 'string') ***REMOVED***
    return new Date(argument)
  ***REMOVED***

  var options = dirtyOptions || ***REMOVED******REMOVED***
  var additionalDigits = options.additionalDigits
  if (additionalDigits == null) ***REMOVED***
    additionalDigits = DEFAULT_ADDITIONAL_DIGITS
  ***REMOVED*** else ***REMOVED***
    additionalDigits = Number(additionalDigits)
  ***REMOVED***

  var dateStrings = splitDateString(argument)

  var parseYearResult = parseYear(dateStrings.date, additionalDigits)
  var year = parseYearResult.year
  var restDateString = parseYearResult.restDateString

  var date = parseDate(restDateString, year)

  if (date) ***REMOVED***
    var timestamp = date.getTime()
    var time = 0
    var offset

    if (dateStrings.time) ***REMOVED***
      time = parseTime(dateStrings.time)
    ***REMOVED***

    if (dateStrings.timezone) ***REMOVED***
      offset = parseTimezone(dateStrings.timezone) * MILLISECONDS_IN_MINUTE
    ***REMOVED*** else ***REMOVED***
      var fullTime = timestamp + time
      var fullTimeDate = new Date(fullTime)

      offset = getTimezoneOffsetInMilliseconds(fullTimeDate)

      // Adjust time when it's coming from DST
      var fullTimeDateNextDay = new Date(fullTime)
      fullTimeDateNextDay.setDate(fullTimeDate.getDate() + 1)
      var offsetDiff =
        getTimezoneOffsetInMilliseconds(fullTimeDateNextDay) -
        getTimezoneOffsetInMilliseconds(fullTimeDate)
      if (offsetDiff > 0) ***REMOVED***
        offset += offsetDiff
      ***REMOVED***
    ***REMOVED***

    return new Date(timestamp + time + offset)
  ***REMOVED*** else ***REMOVED***
    return new Date(argument)
  ***REMOVED***
***REMOVED***

function splitDateString (dateString) ***REMOVED***
  var dateStrings = ***REMOVED******REMOVED***
  var array = dateString.split(parseTokenDateTimeDelimeter)
  var timeString

  if (parseTokenPlainTime.test(array[0])) ***REMOVED***
    dateStrings.date = null
    timeString = array[0]
  ***REMOVED*** else ***REMOVED***
    dateStrings.date = array[0]
    timeString = array[1]
  ***REMOVED***

  if (timeString) ***REMOVED***
    var token = parseTokenTimezone.exec(timeString)
    if (token) ***REMOVED***
      dateStrings.time = timeString.replace(token[1], '')
      dateStrings.timezone = token[1]
    ***REMOVED*** else ***REMOVED***
      dateStrings.time = timeString
    ***REMOVED***
  ***REMOVED***

  return dateStrings
***REMOVED***

function parseYear (dateString, additionalDigits) ***REMOVED***
  var parseTokenYYY = parseTokensYYY[additionalDigits]
  var parseTokenYYYYY = parseTokensYYYYY[additionalDigits]

  var token

  // YYYY or ±YYYYY
  token = parseTokenYYYY.exec(dateString) || parseTokenYYYYY.exec(dateString)
  if (token) ***REMOVED***
    var yearString = token[1]
    return ***REMOVED***
      year: parseInt(yearString, 10),
      restDateString: dateString.slice(yearString.length)
    ***REMOVED***
  ***REMOVED***

  // YY or ±YYY
  token = parseTokenYY.exec(dateString) || parseTokenYYY.exec(dateString)
  if (token) ***REMOVED***
    var centuryString = token[1]
    return ***REMOVED***
      year: parseInt(centuryString, 10) * 100,
      restDateString: dateString.slice(centuryString.length)
    ***REMOVED***
  ***REMOVED***

  // Invalid ISO-formatted year
  return ***REMOVED***
    year: null
  ***REMOVED***
***REMOVED***

function parseDate (dateString, year) ***REMOVED***
  // Invalid ISO-formatted year
  if (year === null) ***REMOVED***
    return null
  ***REMOVED***

  var token
  var date
  var month
  var week

  // YYYY
  if (dateString.length === 0) ***REMOVED***
    date = new Date(0)
    date.setUTCFullYear(year)
    return date
  ***REMOVED***

  // YYYY-MM
  token = parseTokenMM.exec(dateString)
  if (token) ***REMOVED***
    date = new Date(0)
    month = parseInt(token[1], 10) - 1
    date.setUTCFullYear(year, month)
    return date
  ***REMOVED***

  // YYYY-DDD or YYYYDDD
  token = parseTokenDDD.exec(dateString)
  if (token) ***REMOVED***
    date = new Date(0)
    var dayOfYear = parseInt(token[1], 10)
    date.setUTCFullYear(year, 0, dayOfYear)
    return date
  ***REMOVED***

  // YYYY-MM-DD or YYYYMMDD
  token = parseTokenMMDD.exec(dateString)
  if (token) ***REMOVED***
    date = new Date(0)
    month = parseInt(token[1], 10) - 1
    var day = parseInt(token[2], 10)
    date.setUTCFullYear(year, month, day)
    return date
  ***REMOVED***

  // YYYY-Www or YYYYWww
  token = parseTokenWww.exec(dateString)
  if (token) ***REMOVED***
    week = parseInt(token[1], 10) - 1
    return dayOfISOYear(year, week)
  ***REMOVED***

  // YYYY-Www-D or YYYYWwwD
  token = parseTokenWwwD.exec(dateString)
  if (token) ***REMOVED***
    week = parseInt(token[1], 10) - 1
    var dayOfWeek = parseInt(token[2], 10) - 1
    return dayOfISOYear(year, week, dayOfWeek)
  ***REMOVED***

  // Invalid ISO-formatted date
  return null
***REMOVED***

function parseTime (timeString) ***REMOVED***
  var token
  var hours
  var minutes

  // hh
  token = parseTokenHH.exec(timeString)
  if (token) ***REMOVED***
    hours = parseFloat(token[1].replace(',', '.'))
    return (hours % 24) * MILLISECONDS_IN_HOUR
  ***REMOVED***

  // hh:mm or hhmm
  token = parseTokenHHMM.exec(timeString)
  if (token) ***REMOVED***
    hours = parseInt(token[1], 10)
    minutes = parseFloat(token[2].replace(',', '.'))
    return (hours % 24) * MILLISECONDS_IN_HOUR +
      minutes * MILLISECONDS_IN_MINUTE
  ***REMOVED***

  // hh:mm:ss or hhmmss
  token = parseTokenHHMMSS.exec(timeString)
  if (token) ***REMOVED***
    hours = parseInt(token[1], 10)
    minutes = parseInt(token[2], 10)
    var seconds = parseFloat(token[3].replace(',', '.'))
    return (hours % 24) * MILLISECONDS_IN_HOUR +
      minutes * MILLISECONDS_IN_MINUTE +
      seconds * 1000
  ***REMOVED***

  // Invalid ISO-formatted time
  return null
***REMOVED***

function parseTimezone (timezoneString) ***REMOVED***
  var token
  var absoluteOffset

  // Z
  token = parseTokenTimezoneZ.exec(timezoneString)
  if (token) ***REMOVED***
    return 0
  ***REMOVED***

  // ±hh
  token = parseTokenTimezoneHH.exec(timezoneString)
  if (token) ***REMOVED***
    absoluteOffset = parseInt(token[2], 10) * 60
    return (token[1] === '+') ? -absoluteOffset : absoluteOffset
  ***REMOVED***

  // ±hh:mm or ±hhmm
  token = parseTokenTimezoneHHMM.exec(timezoneString)
  if (token) ***REMOVED***
    absoluteOffset = parseInt(token[2], 10) * 60 + parseInt(token[3], 10)
    return (token[1] === '+') ? -absoluteOffset : absoluteOffset
  ***REMOVED***

  return 0
***REMOVED***

function dayOfISOYear (isoYear, week, day) ***REMOVED***
  week = week || 0
  day = day || 0
  var date = new Date(0)
  date.setUTCFullYear(isoYear, 0, 4)
  var fourthOfJanuaryDay = date.getUTCDay() || 7
  var diff = week * 7 + day + 1 - fourthOfJanuaryDay
  date.setUTCDate(date.getUTCDate() + diff)
  return date
***REMOVED***

module.exports = parse
