// supports-color/index.js
'use strict';
const os = require('os');
const tty = require('tty');
const hasFlag = require('has-flag');

const ***REMOVED***env***REMOVED*** = process;

let forceColor;
if (hasFlag('no-color') ||
	hasFlag('no-colors') ||
	hasFlag('color=false') ||
	hasFlag('color=never')) ***REMOVED***
	forceColor = 0;
***REMOVED*** else if (hasFlag('color') ||
	hasFlag('colors') ||
	hasFlag('color=true') ||
	hasFlag('color=always')) ***REMOVED***
	forceColor = 1;
***REMOVED***

if ('FORCE_COLOR' in env) ***REMOVED***
	if (env.FORCE_COLOR === 'true') ***REMOVED***
		forceColor = 1;
	***REMOVED*** else if (env.FORCE_COLOR === 'false') ***REMOVED***
		forceColor = 0;
	***REMOVED*** else ***REMOVED***
		forceColor = env.FORCE_COLOR.length === 0 ? 1 : Math.min(parseInt(env.FORCE_COLOR, 10), 3);
	***REMOVED***
***REMOVED***

function translateLevel(level) ***REMOVED***
	if (level === 0) ***REMOVED***
		return false;
	***REMOVED***

	return ***REMOVED***
		level,
		hasBasic: true,
		has256: level >= 2,
		has16m: level >= 3
	***REMOVED***;
***REMOVED***

function supportsColor(haveStream, streamIsTTY) ***REMOVED***
	if (forceColor === 0) ***REMOVED***
		return 0;
	***REMOVED***

	if (hasFlag('color=16m') ||
		hasFlag('color=full') ||
		hasFlag('color=truecolor')) ***REMOVED***
		return 3;
	***REMOVED***

	if (hasFlag('color=256')) ***REMOVED***
		return 2;
	***REMOVED***

	if (haveStream && !streamIsTTY && forceColor === undefined) ***REMOVED***
		return 0;
	***REMOVED***

	const min = forceColor || 0;

	if (env.TERM === 'dumb') ***REMOVED***
		return min;
	***REMOVED***

	if (process.platform === 'win32') ***REMOVED***
		// Windows 10 build 10586 is the first Windows release that supports 256 colors.
		// Windows 10 build 14931 is the first release that supports 16m/TrueColor.
		const osRelease = os.release().split('.');
		if (
			Number(osRelease[0]) >= 10 &&
			Number(osRelease[2]) >= 10586
		) ***REMOVED***
			return Number(osRelease[2]) >= 14931 ? 3 : 2;
		***REMOVED***

		return 1;
	***REMOVED***

	if ('CI' in env) ***REMOVED***
		if (['TRAVIS', 'CIRCLECI', 'APPVEYOR', 'GITLAB_CI'].some(sign => sign in env) || env.CI_NAME === 'codeship') ***REMOVED***
			return 1;
		***REMOVED***

		return min;
	***REMOVED***

	if ('TEAMCITY_VERSION' in env) ***REMOVED***
		return /^(9\.(0*[1-9]\d*)\.|\d***REMOVED***2,***REMOVED***\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
	***REMOVED***

	if ('GITHUB_ACTIONS' in env) ***REMOVED***
		return 1;
	***REMOVED***

	if (env.COLORTERM === 'truecolor') ***REMOVED***
		return 3;
	***REMOVED***

	if ('TERM_PROGRAM' in env) ***REMOVED***
		const version = parseInt((env.TERM_PROGRAM_VERSION || '').split('.')[0], 10);

		switch (env.TERM_PROGRAM) ***REMOVED***
			case 'iTerm.app':
				return version >= 3 ? 3 : 2;
			case 'Apple_Terminal':
				return 2;
			// No default
		***REMOVED***
	***REMOVED***

	if (/-256(color)?$/i.test(env.TERM)) ***REMOVED***
		return 2;
	***REMOVED***

	if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)) ***REMOVED***
		return 1;
	***REMOVED***

	if ('COLORTERM' in env) ***REMOVED***
		return 1;
	***REMOVED***

	return min;
***REMOVED***

function getSupportLevel(stream) ***REMOVED***
	const level = supportsColor(stream, stream && stream.isTTY);
	return translateLevel(level);
***REMOVED***

module.exports = ***REMOVED***
	supportsColor: getSupportLevel,
	stdout: translateLevel(supportsColor(true, tty.isatty(1))),
	stderr: translateLevel(supportsColor(true, tty.isatty(2)))
***REMOVED***;
