// update-notifier/index.js
'use strict';
const ***REMOVED***spawn***REMOVED*** = require('child_process');
const path = require('path');
const ***REMOVED***format***REMOVED*** = require('util');
const importLazy = require('import-lazy')(require);

const configstore = importLazy('configstore');
const chalk = importLazy('chalk');
const semverDiff = importLazy('semver-diff');
const latestVersion = importLazy('latest-version');
const isNpm = importLazy('is-npm');
const isInstalledGlobally = importLazy('is-installed-globally');
const isYarnGlobal = importLazy('is-yarn-global');
const hasYarn = importLazy('has-yarn');
const boxen = importLazy('boxen');
const xdgBasedir = importLazy('xdg-basedir');
const isCi = importLazy('is-ci');
const pupa = importLazy('pupa');

const ONE_DAY = 1000 * 60 * 60 * 24;

class UpdateNotifier ***REMOVED***
	constructor(options = ***REMOVED******REMOVED***) ***REMOVED***
		this.options = options;
		options.pkg = options.pkg || ***REMOVED******REMOVED***;
		options.distTag = options.distTag || 'latest';

		// Reduce pkg to the essential keys. with fallback to deprecated options
		// TODO: Remove deprecated options at some point far into the future
		options.pkg = ***REMOVED***
			name: options.pkg.name || options.packageName,
			version: options.pkg.version || options.packageVersion
		***REMOVED***;

		if (!options.pkg.name || !options.pkg.version) ***REMOVED***
			throw new Error('pkg.name and pkg.version required');
		***REMOVED***

		this.packageName = options.pkg.name;
		this.packageVersion = options.pkg.version;
		this.updateCheckInterval = typeof options.updateCheckInterval === 'number' ? options.updateCheckInterval : ONE_DAY;
		this.disabled = 'NO_UPDATE_NOTIFIER' in process.env ||
			process.env.NODE_ENV === 'test' ||
			process.argv.includes('--no-update-notifier') ||
			isCi();
		this.shouldNotifyInNpmScript = options.shouldNotifyInNpmScript;

		if (!this.disabled) ***REMOVED***
			try ***REMOVED***
				const ConfigStore = configstore();
				this.config = new ConfigStore(`update-notifier-$***REMOVED***this.packageName***REMOVED***`, ***REMOVED***
					optOut: false,
					// Init with the current time so the first check is only
					// after the set interval, so not to bother users right away
					lastUpdateCheck: Date.now()
				***REMOVED***);
			***REMOVED*** catch (_) ***REMOVED***
				// Expecting error code EACCES or EPERM
				const message =
					chalk().yellow(format(' %s update check failed ', options.pkg.name)) +
					format('\n Try running with %s or get access ', chalk().cyan('sudo')) +
					'\n to the local update config store via \n' +
					chalk().cyan(format(' sudo chown -R $USER:$(id -gn $USER) %s ', xdgBasedir().config));

				process.on('exit', () => ***REMOVED***
					console.error('\n' + boxen()(message, ***REMOVED***align: 'center'***REMOVED***));
				***REMOVED***);
			***REMOVED***
		***REMOVED***
	***REMOVED***

	check() ***REMOVED***
		if (
			!this.config ||
			this.config.get('optOut') ||
			this.disabled
		) ***REMOVED***
			return;
		***REMOVED***

		this.update = this.config.get('update');

		if (this.update) ***REMOVED***
			// Use the real latest version instead of the cached one
			this.update.current = this.packageVersion;

			// Clear cached information
			this.config.delete('update');
		***REMOVED***

		// Only check for updates on a set interval
		if (Date.now() - this.config.get('lastUpdateCheck') < this.updateCheckInterval) ***REMOVED***
			return;
		***REMOVED***

		// Spawn a detached process, passing the options as an environment property
		spawn(process.execPath, [path.join(__dirname, 'check.js'), JSON.stringify(this.options)], ***REMOVED***
			detached: true,
			stdio: 'ignore'
		***REMOVED***).unref();
	***REMOVED***

	async fetchInfo() ***REMOVED***
		const ***REMOVED***distTag***REMOVED*** = this.options;
		const latest = await latestVersion()(this.packageName, ***REMOVED***version: distTag***REMOVED***);

		return ***REMOVED***
			latest,
			current: this.packageVersion,
			type: semverDiff()(this.packageVersion, latest) || distTag,
			name: this.packageName
		***REMOVED***;
	***REMOVED***

	notify(options) ***REMOVED***
		const suppressForNpm = !this.shouldNotifyInNpmScript && isNpm().isNpmOrYarn;
		if (!process.stdout.isTTY || suppressForNpm || !this.update || this.update.current === this.update.latest) ***REMOVED***
			return this;
		***REMOVED***

		options = Object.assign(***REMOVED***
			isGlobal: isInstalledGlobally(),
			isYarnGlobal: isYarnGlobal()()
		***REMOVED***, options);

		let installCommand;

		if (options.isYarnGlobal) ***REMOVED***
			installCommand = `yarn global add $***REMOVED***this.packageName***REMOVED***`;
		***REMOVED*** else if (options.isGlobal) ***REMOVED***
			installCommand = `npm i -g $***REMOVED***this.packageName***REMOVED***`;
		***REMOVED*** else if (hasYarn()()) ***REMOVED***
			installCommand = `yarn add $***REMOVED***this.packageName***REMOVED***`;
		***REMOVED*** else ***REMOVED***
			installCommand = `npm i $***REMOVED***this.packageName***REMOVED***`;
		***REMOVED***

		const defaultTemplate = 'Update available ' +
			chalk().dim('***REMOVED***currentVersion***REMOVED***') +
			chalk().reset(' → ') +
			chalk().green('***REMOVED***latestVersion***REMOVED***') +
			' \nRun ' + chalk().cyan('***REMOVED***updateCommand***REMOVED***') + ' to update';

		const template = options.message || defaultTemplate;

		options.boxenOptions = options.boxenOptions || ***REMOVED***
			padding: 1,
			margin: 1,
			align: 'center',
			borderColor: 'yellow',
			borderStyle: 'round'
		***REMOVED***;

		const message = '\n' + boxen()(
			pupa()(template, ***REMOVED***
				packageName: this.packageName,
				currentVersion: this.update.current,
				latestVersion: this.update.latest,
				updateCommand: installCommand
			***REMOVED***),
			options.boxenOptions
		);

		if (options.defer === false) ***REMOVED***
			console.error(message);
		***REMOVED*** else ***REMOVED***
			process.on('exit', () => ***REMOVED***
				console.error(message);
			***REMOVED***);

			process.on('SIGINT', () => ***REMOVED***
				console.error('');
				process.exit();
			***REMOVED***);
		***REMOVED***

		return this;
	***REMOVED***
***REMOVED***

module.exports = options => ***REMOVED***
	const updateNotifier = new UpdateNotifier(options);
	updateNotifier.check();
	return updateNotifier;
***REMOVED***;

module.exports.UpdateNotifier = UpdateNotifier;
